### PR TITLE
877 undefined method GitHub url airbrake

### LINKED ISSUE #877: [Airbrake] [Production] undefined method `[]' for nil:NilClass
Check full details on Airbrake: [#1637042363901145721](https://herokuapp21498474herokucom.airbrake.io/projects/117119/groups/1637042363901145721)

Airbrake project: **websiteone-production**

Backtrace:
```
[PROJECT_ROOT]/app/models/project.rb:53:in github_repo_name
[PROJECT_ROOT]/app/jobs/github_commits_job.rb:23:in github_url
[PROJECT_ROOT]/app/jobs/github_commits_job.rb:16:in update_total_commit_count_for
[PROJECT_ROOT]/app/jobs/github_commits_job.rb:8:in block in run
[PROJECT_ROOT]/vendor/bundle/ruby/2.2.0/gems/activerecord-4.1.6/lib/active_record/relation/delegation.rb:46:in each
[PROJECT_ROOT]/vendor/bundle/ruby/2.2.0/gems/activerecord-4.1.6/lib/active_record/relation/delegation.rb:46:in each
[PROJECT_ROOT]/app/jobs/github_commits_job.rb:7:in run
[PROJECT_ROOT]/lib/tasks/scheduler.rake:4:in block in <top (required)>
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:240:in call
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:240:in block in execute
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:235:in each
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:235:in execute
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:179:in block in invoke_with_call_chain
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/monitor.rb:211:in mon_synchronize
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:172:in invoke_with_call_chain
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/task.rb:165:in invoke
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:150:in invoke_task
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:106:in block (2 levels) in top_level
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:106:in each
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:106:in block in top_level
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:115:in run_with_threads
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:100:in top_level
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:78:in block in run
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:176:in standard_exception_handling
[PROJECT_ROOT]/vendor/ruby-2.2.2/lib/ruby/2.2.0/rake/application.rb:75:in run
[PROJECT_ROOT]/bin/rake:4:in <main>
```

### PR DESCRIPTION
**Pull request by touchps**
Fixes #877

### CODE REVIEW

File: features/devops/github_commits.feature [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Feature: Update github stats of all projects
```
Comment (tansaku) [inline features/devops/github_commits.feature]: you need to add @vcr here

**Comment by touchps** [comment]
To be honest i'm not satisfied with the feature as it passes without putting any fix. Also i tried running 
`bundle exec rake fetc_github_commits` with projects having github_url as null and it ran fine.

Clearly the code works fine with projects having null entries for github_url.
I tested with repo that doesn't exist and it gave entirely different error than the one in airbrake.

**Comment by touchps** [comment]
Found a way to reproduce. Its the incomplete url that causes the issue. For example it gives similar errors as airbrake with [https://github.com/edx] (listed as last project).

Perhaps wrapping in rescue block will be a better solution.

**Comment by tansaku** [comment]
@touchps good progress on this - however I think we need the test to fail in order to expose the error right?

**Comment by touchps** [comment]
Yes, the cucumber feature was passing initially because we didn't had any appropriate test cases for which airbrake issue was failing.

**Comment by touchps** [comment]
Few notes:

- I wanted to put one more test in 'github_commits_job_spec.rb` for the failed url, but it resulted in series of errors rather than failing tests. Still under consideration.

- Including exception check at higher level made more sense based on the fact that we should throw exceptions as early when found but catch them as late as possible. Also allowing exception to bubble up from project model made more sense as it may or may not affect other dependent code.

File: app/jobs/github_commits_job.rb [hunk 2]
```diff
@@ -5,8 +5,12 @@ module GithubCommitsJob
 
   def run
     Project.with_github_url.each do |project|
-      update_total_commit_count_for(project)
-      update_user_commit_counts_for(project)
+      begin    
+        update_total_commit_count_for(project)
+        update_user_commit_counts_for(project)
+      rescue Exception
```
Comment (tansaku) [inline app/jobs/github_commits_job.rb]: should we rescue specifically the exception we get?  maybe not ...

File: app/jobs/github_commits_job.rb [hunk 3]
```diff
@@ -5,8 +5,12 @@ module GithubCommitsJob
...
+        Rails.logger.warn "#{project.github_url} may have caused the issue. Commit update terminated for this project!"
```
Comment (tansaku) [inline app/jobs/github_commits_job.rb]: good use of the logger

File: features/step_definitions/projects_steps.rb [hunk 4]
```diff
@@ -110,3 +110,13 @@
 
 end
 
+Then(/^I Should see projects with following details:$/) do |table|
```
Comment (tansaku) [inline features/step_definitions/projects_steps.rb]: Should --> should

File: features/devops/github_commits.feature [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      | WebsiteTwo   | awesome autograder  | https://github.com/AgileVentures/WebsiteTwo   | active   | 1            |
+      | WebsiteOne   | website one project | https://github.com/AgileVentures/WebsiteOne   | inactive | 1            |
+      | edx          | MOOC platform       | https://github.com/edx                        | active   | 1            |
+      | Unity        | Unity project       |                                               | active   | 1            |
+      | LocalSupport | Scheduler queue     | https://github.com/AgileVentures/LocalSupport | active   | 1            |
+
+   Scenario: Update github commit count of all projects with valid github_url
+      When I run rake task "fetch_github_commits"
+      Then I Should see projects with following details:
+      | title        | description         | github_url                                    | status   | commit_count |
```
Comment (tansaku) [inline features/devops/github_commits.feature]: could remove description column here

File: spec/models/project_spec.rb [hunk 6]
```diff
@@ -33,6 +33,11 @@
       expect(subject).to_not be_valid
     end
 
+    it 'should throw error for incomplete github url' do
+      subject.github_url = 'https://github.com/edx'
+      expect{ subject.github_repo_name }.to raise_error
```
Comment (tansaku) [inline spec/models/project_spec.rb]: should avoid raise_error without specifics - ideally add exception type or error message

**Comment by tansaku** [comment]
@diraulo I think this is good to go - great work @touchps - @diraulo feel free to merge unless you see any issues

**Comment by diraulo** [comment]
great work on this @touchps

**Pull request merged**